File: skypydb/database/mixins/vector/add_items.py
"""Insert operations for vector collection records."""

import json
from datetime import datetime
from typing import Any, Dict, List, Optional

from skypydb.security.validation import InputValidator


class AddItemsMixin:
    """Provide record insertion for vector collections."""

    def add(
        self,
        collection_name: str,
        ids: List[str],
        embeddings: Optional[List[List[float]]] = None,
        documents: Optional[List[str]] = None,
        metadatas: Optional[List[Dict[str, Any]]] = None,
    ) -> List[str]:
        """Insert records into a collection by IDs and vectors/documents."""

        collection_name = InputValidator.validate_table_name(collection_name)
        if not self.collection_exists(collection_name):
            raise ValueError(f"Collection '{collection_name}' not found")
        if embeddings is None and documents is None:
            raise ValueError("Either embeddings or documents must be provided")

        if embeddings is None:
            if self.embedding_function is None:
                raise ValueError(
                    "Documents provided but no embedding function set. "
                    "Either provide embeddings directly or set an embedding_function."
                )
            embeddings = self.embedding_function(documents or [])

        n_items = len(ids)
        if len(embeddings) != n_items:
            raise ValueError(
                f"Number of embeddings ({len(embeddings)}) doesn't match number of IDs ({n_items})"
            )
        if documents is not None and len(documents) != n_items:
            raise ValueError(
                f"Number of documents ({len(documents)}) doesn't match number of IDs ({n_items})"
            )
        if metadatas is not None and len(metadatas) != n_items:
            raise ValueError(
                f"Number of metadatas ({len(metadatas)}) doesn't match number of IDs ({n_items})"
            )

        cursor = self.conn.cursor()
        now = datetime.now().isoformat()

        for idx, item_id in enumerate(ids):
            cursor.execute(
                f"""
                INSERT OR REPLACE INTO [vec_{collection_name}]
                (id, document, embedding, metadata, created_at)
                VALUES (?, ?, ?, ?, ?)
                """,
                (
                    item_id,
                    documents[idx] if documents else None,
                    json.dumps(embeddings[idx]),
                    json.dumps(metadatas[idx]) if metadatas else None,
                    now,
                ),
            )
        self.conn.commit()
        return ids

File: skypydb-js/src/database/mixins/vector/add_items.ts
import { CollectionNotFoundError } from "../../../errors";
import { InputValidator } from "../../../security/validation";
import type { EmbeddingMatrix, Metadata } from "../../../types";
import { EmbeddingFunctionMixin } from "./embedding_function";

export abstract class AddItemsMixin extends EmbeddingFunctionMixin {
  async add(
    collection_name: string,
    ids: string[],
    embeddings?: EmbeddingMatrix,
    documents?: string[],
    metadatas?: Metadata[]
  ): Promise<string[]> {
    const validated = InputValidator.validate_table_name(collection_name);
    if (!this.collection_exists(validated)) {
      throw new CollectionNotFoundError(`Collection '${validated}' not found`);
    }
    if (!embeddings && !documents) {
      throw new Error("Either embeddings or documents must be provided");
    }

    let resolved_embeddings = embeddings;
    if (!resolved_embeddings) {
      if (!this.embedding_function) {
        throw new Error(
          "Documents provided but no embedding function set. Either provide embeddings directly or set an embedding_function."
        );
      }
      resolved_embeddings = await this.embedding_function(documents ?? []);
    }

    const n_items = ids.length;
    if (resolved_embeddings.length !== n_items) {
      throw new Error(
        `Number of embeddings (${resolved_embeddings.length}) doesn't match number of IDs (${n_items})`
      );
    }
    if (documents && documents.length !== n_items) {
      throw new Error(
        `Number of documents (${documents.length}) doesn't match number of IDs (${n_items})`
      );
    }
    if (metadatas && metadatas.length !== n_items) {
      throw new Error(
        `Number of metadatas (${metadatas.length}) doesn't match number of IDs (${n_items})`
      );
    }

    const statement = this.conn.prepare(
      `
      INSERT OR REPLACE INTO [vec_${validated}]
      (id, document, embedding, metadata, created_at)
      VALUES (?, ?, ?, ?, ?)
      `
    );

    const now = new Date().toISOString();
    for (let index = 0; index < n_items; index += 1) {
      statement.run(
        ids[index],
        documents ? documents[index] : null,
        JSON.stringify(resolved_embeddings[index]),
        metadatas ? JSON.stringify(metadatas[index]) : null,
        now
      );
    }

    return ids;
  }
}
